### ISSUE TITLE
Representing a scene graph using StellarDiGraph for caption generation (GCN-LSTM)

### ISSUE THREAD

**Issue by medea-learner** [stellargraph/stellargraph#1775]
Hi,
In _https://community.stellargraph.io/t/use-gcn-lstm-to-generate-descriptions-for-scenes-represented-by-graphs/170/5_ , the questioner has given an example of what his model will receive in input and what will predict in output. I want to build a model like this model, but what problem I encountered is how to represent that type of graph with **StellarDiGraph** (keep in mind that I tried one representation but my model didn't learn from (i.e., both train accuracy and val accuracy remain almost the same whatever the number of epochs and the layer used and ...).
Have anyone an idea of how to represent that type of graphs with **StellarDiGraph** to be able to learn from?
Thanks for advance.

**Comment by huonw** [comment]
Hi, thanks for getting in touch.

As suggested on the linked thread, whole-graph summarisation is best served using a graph classification model to encode the graph to a vector, and then generating sequences from the vector. GCN-LSTM is designed for prediction on a graph with a time-series associated with each node. See https://stellargraph.readthedocs.io/en/stable/demos/graph-classification/index.html for info about graph classification.

Just to make sure we're on the same page: are you trying to use GCN-LSTM with a direct graph/`StellarDiGraph`, or are you using one of the graph classification models?

> keep in mind that I tried one representation but my model didn't learn from (i.e., both train accuracy and val accuracy remain almost the same whatever the number of epochs and the layer used and ...

We'll be able to be much more helpful if you give us a little more info about this. It sounds like you tried something but it failed to learn? Could you say more about what it was? For instance, maybe there's a misconfiguration in the model.

**Comment by medea-learner** [comment]
Hi @huonw, thank you for yr response,
What I did exactly is to represent the scenes by a direct graphs (in that case _**StellarDiGraph**_) and then pass these graphs into the following **_GCN-LSTM_** model :

    gc_model = GCNSupervisedGraphClassification(
        layer_sizes=[64, 64],
        activations=["relu", "relu"],
        generator=generator,
        dropout=0.5,
    )
    x_inp, x_out = gc_model.in_out_tensors()
    predictions = tf.keras.layers.Dense(units=64, activation="relu")(x_out)
    predictions = tf.keras.layers.Dense(units=32, activation="linear")(predictions)
    predictions = tf.keras.layers.Embedding(t_vocabsize, EMBED_SIZE, input_length=32)(predictions)
    predictions = tf.keras.layers.SpatialDropout1D(0.2)(predictions)
    predictions = tf.keras.layers.Bidirectional(tf.compat.v1.keras.layers.CuDNNLSTM(HIDDEN_SIZE))(predictions)
    predictions = tf.keras.layers.RepeatVector(NB_MAX_OF_WORDS)(predictions)
    predictions = tf.keras.layers.Bidirectional(tf.compat.v1.keras.layers.CuDNNLSTM(HIDDEN_SIZE, return_sequences=True))(predictions)
    predictions = tf.keras.layers.TimeDistributed(tf.keras.layers.Dense(t_vocabsize))(predictions)
    predictions = tf.keras.layers.Activation(ACTIVATION)(predictions)
    
    model = Model(inputs=x_inp, outputs=predictions)
    model.compile(loss="categorical_crossentropy", optimizer=OPTIMIZER, metrics=["accuracy"])

NB : I tried both _**DGCNN**_ and _**GCNSupervisedGraphClassification**_, the result is the same.
> Just to make sure we're on the same page: are you trying to use GCN-LSTM with a direct graph/StellarDiGraph, or are you using one of the graph classification models?

  For the first question, If you mean by _**GCN-LSTM**_ Whose given in the StellarGraph documentation, I tried to encode the graphs into vectors using one of the graph classification model and then pass them to an LSTM model (so, I'm not using the GCN-LSTM model given in the StellarGraph documentation)
  For the second question, I think that the first one gives an answer of that question.

Just to give more clarification of how I represented the graphs using **_StellarDiGraph_** class:

-  each object that I have in the dataset that I'm using has an id.
-  for a given graph, their nodes are the ids of the objects and their(i.e, the nodes) features are one-hot encoded vectors.
-  for the edges features, I could not manage to represent them.

Thanks in advance for spending time to solve my problem.

**Comment by huonw** [comment]
Thanks for all the info. 

It looks like a relatively complicated model, so I'd recommend that we simplify progressively to understand where the failure is. In particular, we can try to narrow down whether there's no signal to learn from, or (more likely) which part of the model is causing issues. Here's some ideas:

1. **Is it the directness?** Use a non-directed `StellarGraph`, to see whether the directed edges are causing the problems (this will of course lose information, there would hopefully still be _something_ in the dataset)
2. **Is it the connection between the graph classification model and the later layers?** Look at the output of the graph classification model separately; one way to do this would be to [train a graph classification model unsupervised](https://stellargraph.readthedocs.io/en/stable/demos/embeddings/gcn-unsupervised-graph-embeddings.html) and use the resulting embeddings in two ways:
   1. look at the embeddings themselves, to understand if there's any signal in them
   2. train a classifier/LSTM on top of those embeddings, to work out if it's the end-to-end training that's causing a problem (if this approach does work better, one could also try semi-supervised training, where the graph classification model is first trained without supervision and then the full LSTM model is fine-tuned end-to-end, similar to https://stellargraph.readthedocs.io/en/stable/demos/node-classification/gcn-deep-graph-infomax-fine-tuning-node-classification.html )
3. **Is it the graph classification layer?** Use something other than the graph classification layer; for instance, ignore the edges and just pool the node features for each graph (with a manual collection of layers)
4. **Is it the other layers?** Simplify the layers after the graph classification, to work out if the graph classification layer is capturing the signal appropriately and that's getting lost in the later layers; for instance:
   1. do binary classification (for instance, "does the caption contain word X?" for some particular `X`) to work out whether there's much signal in the data at all, e.g. instead of the layers from `Embedding`, just do `predictions = Dense(units=1, activation="sigmoid")(predictions)` on the output of the 32-unit dense layer
   2. generalising that, generate a single word caption, like `Dense(units=t_vocabsize, activation="softmax")`
   3. if all of these work, then I guess one would have to progressively re-introduce the `Embedding`, `SpatialDropout1D`, ... `Activation` layers to work out where the failure is

What do you think?

> I'm not using the GCN-LSTM model given in the StellarGraph documentation

Ah, okay, most of my comment was thus just a misunderstanding about what you were referring to. We're on the same page now. 👍
